# Rohan Shah

---

Project - 1

- To download data, go to this link and click download from data tab. You need to have Kaggle account for this.
https://www.kaggle.com/jboysen/us-perm-visas

Instructions for deliverables:
- Do not change anything in any other cell except in cells with comment "# YOUR CODE HERE" or "YOUR ANSWER HERE". If you need to play around with data, play around with it in another notebook. But donot submit that extra work as part of this notebook.
- Do not rename this notebook. Submit file as project1.ipynb only.
- Submit only one notebook. Before submitting, make sure everything runs as expected. To check that, restart the kernel (in the menubar, select Kernel > Restart) and then run all cells (in the menubar, select Cell > Run All).

Instructions for coding:
- Put the dataset in same folder as this file.
- Do not change the name of the dataset. Keep it as "us_perm_visas.csv" only.
- Use groupby in almost all questions.
- Use column 'case_number' for counting purposes. This means if this column has missing value, we don't count that row even if other columns have values in them.
- If same data is avaialble in multiple columns, use column with less missing values.
- Ignore missing data.
- Your output should show exactly what has been asked. Do not show whole big table with all the information. e.g. if it has been asked to list top 3 countries, your out put should have only 3 countries and not list of all countries.

Hint : All questions can be answered by using only one line of code. So try to use one line of code only to get answer.

In [1]:
import warnings
warnings.simplefilter('ignore')
import pandas as pd
data = pd.read_csv("us_perm_visas.csv")

In [2]:
data.head()

add_these_pw_job_title_9089 agent_city agent_firm_name agent_state  \
0                         NaN        NaN             NaN         NaN   
1                         NaN        NaN             NaN         NaN   
2                         NaN        NaN             NaN         NaN   
3                         NaN        NaN             NaN         NaN   
4                         NaN        NaN             NaN         NaN   

  application_type        case_no case_number case_received_date case_status  \
0             PERM  A-07323-97014         NaN                NaN   Certified   
1             PERM  A-07332-99439         NaN                NaN      Denied   
2             PERM  A-07333-99643         NaN                NaN   Certified   
3             PERM  A-07339-01930         NaN                NaN   Certified   
4             PERM  A-07345-03565         NaN                NaN   Certified   

  class_of_admission              ...               ri_pvt_employment_firm_to  \
0                J-1              ...                                     NaN   
1                B-2              ...                                     NaN   
2               H-1B              ...                                     NaN   
3                B-2              ...                                     NaN   
4                L-1              ...                                     NaN   

  ri_us_workers_considered schd_a_sheepherder     us_economic_sector  \
0                      NaN                NaN                     IT   
1                      NaN                NaN  Other Economic Sector   
2                      NaN                NaN              Aerospace   
3                      NaN                NaN  Other Economic Sector   
4                      NaN                NaN           Advanced Mfg   

  wage_offer_from_9089 wage_offer_to_9089 wage_offer_unit_of_pay_9089  \
0                75629                NaN                          yr   
1                37024                NaN                          yr   
2                47923                NaN                          yr   
3                10.97                NaN                          hr   
4               100000                NaN                          yr   

  wage_offered_from_9089 wage_offered_to_9089  wage_offered_unit_of_pay_9089  
0                    NaN                  NaN                            NaN  
1                    NaN                  NaN                            NaN  
2                    NaN                  NaN                            NaN  
3                    NaN                  NaN                            NaN  
4                    NaN                  NaN                            NaN  

[5 rows x 154 columns]

In [3]:
data.shape

(374362, 154)

In [4]:
for i in data.columns:
    print(i)

add_these_pw_job_title_9089
agent_city
agent_firm_name
agent_state
application_type
case_no
case_number
case_received_date
case_status
class_of_admission
country_of_citizenship
country_of_citzenship
decision_date
employer_address_1
employer_address_2
employer_city
employer_country
employer_decl_info_title
employer_name
employer_num_employees
employer_phone
employer_phone_ext
employer_postal_code
employer_state
employer_yr_estab
foreign_worker_info_alt_edu_experience
foreign_worker_info_birth_country
foreign_worker_info_city
foreign_worker_info_education
foreign_worker_info_education_other
foreign_worker_info_inst
foreign_worker_info_major
foreign_worker_info_postal_code
foreign_worker_info_rel_occup_exp
foreign_worker_info_req_experience
foreign_worker_info_state
foreign_worker_info_training_comp
foreign_worker_ownership_interest
foreign_worker_yr_rel_edu_completed
fw_info_alt_edu_experience
fw_info_birth_country
fw_info_education_other
fw_info_postal_code
fw_info_rel_occup_exp
fw_info

Q1. What is the count of applications submitted by each country?

In [5]:
data['country_citizenship'] =data['country_of_citizenship'].fillna('').astype(str)+data['country_of_citzenship'].fillna('').astype(str)
data.groupby(['country_citizenship']).size().sort_values(ascending=False).rename('No_of_application').reset_index()

country_citizenship  No_of_application
0                       INDIA             205158
1                       CHINA              28861
2                 SOUTH KOREA              24761
3                      CANADA              14804
4                      MEXICO               8961
5                 PHILIPPINES               8631
6              UNITED KINGDOM               4972
7                      TAIWAN               4173
8                    PAKISTAN               3962
9                      FRANCE               3173
10                      NEPAL               3047
11                      JAPAN               2970
12                     BRAZIL               2919
13                  VENEZUELA               2738
14                     TURKEY               2551
15                     RUSSIA               2524
16                    GERMANY               2364
17                       IRAN               2356
18                    VIETNAM               2290
19                      ITALY               1948
20                     ISRAEL               1914
21                   COLOMBIA               1843
22                  AUSTRALIA               1696
23                    UKRAINE               1502
24                      SPAIN               1447
25                     POLAND               1424
26                  ARGENTINA               1189
27                    NIGERIA               1147
28                 BANGLADESH               1098
29                   MALAYSIA               1059
..                        ...                ...
173               NORTH KOREA                  4
174                    BRUNEI                  3
175             GUINEA-BISSAU                  3
176                SEYCHELLES                  3
177                MOZAMBIQUE                  3
178            CAYMAN ISLANDS                  3
179                   VANUATU                  3
180  TURKS AND CAICOS ISLANDS                  3
181     SAO TOME AND PRINCIPE                  2
182    BRITISH VIRGIN ISLANDS                  2
183         REPUBLIC OF CONGO                  2
184               SOUTH SUDAN                  2
185                CAPE VERDE                  2
186  CENTRAL AFRICAN REPUBLIC                  2
187             LIECHTENSTEIN                  1
188                    MONACO                  1
189      NETHERLANDS ANTILLES                  1
190                   ANDORRA                  1
191         EQUATORIAL GUINEA                  1
192          PAPUA NEW GUINEA                  1
193                     QATAR                  1
194                     SAMOA                  1
195                     ARUBA                  1
196              SOVIET UNION                  1
197                      CUBA                  1
198                   COMOROS                  1
199              SINT MAARTEN                  1
200                      CHAD                  1
201                   SOMALIA                  1
202                  KIRIBATI                  1

[203 rows x 2 columns]

Q2. List top 3 countries with highest number of applications and also show number of applications by them.

In [6]:
data.groupby(['country_citizenship']).size().sort_values(ascending=False).head(3).rename('Applications').reset_index()

country_citizenship  Applications
0               INDIA        205158
1               CHINA         28861
2         SOUTH KOREA         24761

Q3. How many applications were denied?

In [7]:
len(data[data['case_status']=='Denied'])

25649

Q4. How many applicants holding H-1B visas were certified for Permenant Residency? (Consider all forms used for writing "H1-B" in data. Consider only H1-B, not H1-B1)

In [8]:
len(data[(data.case_status=='Certified')&((data.class_of_admission=='H-1B')|(data.class_of_admission=='H1B')|(data.class_of_admission=='AOS/H-1B'))])

139314

Q5. Sort columns in descending order according to number of missing values in them.

In [9]:
data.isnull().sum().sort_values(ascending=False).rename('missing_val').reset_index()

index  missing_val
0                         orig_file_date       374081
1                           orig_case_no       374068
2                  recr_info_job_fair_to       372569
3                recr_info_job_fair_from       372564
4            recr_info_on_campus_recr_to       372497
5          recr_info_on_campus_recr_from       372490
6            ji_live_in_dom_svc_contract       371964
7    foreign_worker_info_education_other       370422
8                job_info_training_field       369883
9           job_info_training_num_months       369583
10             ri_coll_teach_select_date       368632
11             ri_coll_tch_basic_process       367349
12        recr_info_coll_teach_comp_proc       367341
13              job_info_education_other       366593
14           job_info_alt_combo_ed_other       365148
15           recr_info_pro_org_advert_to       360292
16         recr_info_pro_org_advert_from       360286
17             pw_source_name_other_9089       358603
18             ri_pvt_employment_firm_to       357922
19           ri_pvt_employment_firm_from       357920
20              ri_us_workers_considered       357129
21                   naics_2007_us_title       354462
22            recr_info_radio_tv_ad_from       354009
23              recr_info_radio_tv_ad_to       354008
24                    naics_2007_us_code       353846
25                 country_of_citzenship       353788
26                    employer_phone_ext       350901
27          recr_info_prof_org_advert_to       346276
28        recr_info_prof_org_advert_from       346269
29                ri_campus_placement_to       344369
..                                   ...          ...
125           recr_info_professional_occ       135286
126            job_info_foreign_lang_req       135286
127                    job_info_training       135284
128              job_info_job_req_normal       135283
129          ji_live_in_domestic_service       135282
130                   job_info_education       135282
131                  job_info_experience       135281
132            job_info_combo_occupation       135279
133                   case_received_date       135271
134                          case_number       135269
135          wage_offer_unit_of_pay_9089       115916
136                 wage_offer_from_9089       114771
137                        pw_level_9089        27627
138                   class_of_admission        22845
139               country_of_citizenship        20633
140                         pw_soc_title         2336
141                       pw_amount_9089         2216
142                  pw_source_name_9089         2099
143                  pw_unit_of_pay_9089         1572
144                          pw_soc_code          397
145                  job_info_work_state          103
146                   job_info_work_city          102
147                   employer_address_1           42
148                       employer_state           42
149                 employer_postal_code           37
150                        employer_city           14
151                        employer_name           12
152                        decision_date            0
153                          case_status            0
154                  country_citizenship            0

[155 rows x 2 columns]

Q6. List top 10 cities from which employers have submitted the most applications. How many applicarions were submitted by them?

In [10]:
data.groupby(['employer_city']).size().sort_values(ascending=False).head(10).rename('no_of_application').reset_index()

employer_city  no_of_application
0         NEW YORK              16730
1  COLLEGE STATION              12118
2      SANTA CLARA              11113
3         SAN JOSE               9349
4          REDMOND               8804
5    MOUNTAIN VIEW               8259
6          HOUSTON               6446
7        SUNNYVALE               6285
8    SAN FRANCISCO               6254
9            PLANO               5836

Q7. Which are the top 10 employers who have submitted the highest number of applications? How many?

In [11]:
data.groupby(['employer_name']).size().sort_values(ascending=False).head(10).rename('no_of_application').reset_index()

employer_name  no_of_application
0  COGNIZANT TECHNOLOGY SOLUTIONS US CORPORATION              12081
1                          MICROSOFT CORPORATION               8369
2                              INTEL CORPORATION               6564
3                                    GOOGLE INC.               5165
4                           AMAZON CORPORATE LLC               3719
5                                   INFOSYS LTD.               2717
6                            CISCO SYSTEMS, INC.               2634
7                           ORACLE AMERICA, INC.               2584
8                                     APPLE INC.               2583
9                     QUALCOMM TECHNOLOGIES INC.               1710

Q8. What is the average time of processing the applications?

In [12]:
data['processing_time']=pd.to_datetime(data['decision_date'])-pd.to_datetime(data['case_received_date'])
data['processing_time'].dropna().mean()

Timedelta('208 days 01:59:13.654466')

Q9. What is count of applications for each case status(e.g. Certified, Denied etc) for applicants with "Doctorate" degree?

In [13]:
data[data.foreign_worker_info_education=='Doctorate'].groupby('case_status').size().rename('count').reset_index()

case_status  count
0          Certified   5965
1  Certified-Expired   5449
2             Denied    345
3          Withdrawn    484

Q.10 What are the top 3 economic sectors that has maximum applicants? How many applications are submitted from these sectors? (Use case_no column here, instead of case_number)

In [14]:
data.groupby('us_economic_sector').size().rename('No_of_Applications').reset_index().sort_values('No_of_Applications',ascending=False).head(3)

us_economic_sector  No_of_Applications
13                     IT               52617
0            Advanced Mfg               18749
14  Other Economic Sector               15798

Q.11 What is the mean wage of all the certified visa applications?

In [15]:
data['wage']=data['wage_offer_from_9089'].fillna('0').astype(str)+data['wage_offered_from_9089'].fillna('0').astype(str)
data['unit']=data['wage_offer_unit_of_pay_9089'].fillna('').astype(str)+data['wage_offered_unit_of_pay_9089'].fillna('').astype(str)
data["wage"] = data["wage"].convert_objects(convert_numeric=True)
data[data['case_status']=='Certified'].wage.mean()
    

95424.50898910762

Q.12 Show one instresting finding of your own which is not covered above. Show the code and write your finding as comment.

In [16]:
'''
what if we want to know which Specific company has the highest number of application from 
the same branch,also we want the number of employees in that branch
'''
data.groupby(['employer_name','employer_postal_code','employer_state','employer_num_employees']).size().sort_values(ascending=False).rename('no_of_applications').reset_index().head(1)

employer_name employer_postal_code  \
0  COGNIZANT TECHNOLOGY SOLUTIONS US CORPORATION                77845   

  employer_state  employer_num_employees  no_of_applications  
0          TEXAS                 29000.0                4499

Q.13 Write a report on your analysis of this dataset. (Around 200 words, but quality of your report will be taken into account more than count of words)

# Analysis Report

Working with the dataset "us_perm_visas.csv" is quite a challenge as there are many columns in the dataset that have same type of data in them but have missing values inside them; to get rid of this problem we can merge those two columns into one to find the best result possible. 

From the analysis of this dataset we can say that INDIA, CHINA and SOUTH KOREA dominates in terms of number of applicants, who have applied for the visa. From the total visa applicants there were 6.85% of applicants that were denied the visa. From, the total H1-B visa applicants nearly 49% of applicants were given permanent visas. We can see from the analysis that there are many fields that has missing values inside them.  The original file date column has the highest number of missing values inside it. We can also see that top 10 cities combined has 24.35% of the total number of applications. Top 10 companies combined has more than 1/8th percentage (12.85%) of share from total applications. The average processing time for the application is nearly 208 days. From all the applicants with doctorate degree nearly 49% of them were certified, 44% of them have their visa expired and rest of them were denied or withdrawn their applications. IT has the maximum number of applicants among all the other sectors. The mean wage for all the certified visa applicants is 95424.50 USD. COGNIZANT from Texas has the highest number of application among any other company and has total of 29000 employees with 4499 no visa applications
